src/network/ClientConnectionManager.ts

```diff
@@ -0,0 +1,804 @@
+            clientMessage = ClientAuthenticationCodec.encodeRequest(clusterName, null, null, this.clientUuid,
+                CLIENT_TYPE, SERIALIZATION_VERSION, clientVersion, clientName, this.labels);
+        }
+
+        return clientMessage;
+    }
+
+    private checkPartitionCount(newPartitionCount: number): void {
+        const partitionService = this.client.getPartitionService();
+        if (!partitionService.checkAndSetPartitionCount(newPartitionCount)) {
+            throw new ClientNotAllowedInClusterError('Client can not work with this cluster because it has a different ' +
+                'partition count. Expected partition count: ' + partitionService.getPartitionCount() +
+                ', member partition count: ' + newPartitionCount);
+        }
+    }
+
+    private initializeClientOnCluster(targetClusterId: UUID): Promise<void> {
+        if (!targetClusterId.equals(this.clusterId)) {
+            this.logger.warn('ConnectionManager', 'Won\'t send client state to cluster: ' + targetClusterId +
+                ' because switched to a new cluster: ' + this.clusterId);
+            return;
+        }
+
+        return this.client.sendStateToCluster()
+            .then(() => {
+                if (targetClusterId.equals(this.clusterId)) {
+                    this.logger.trace('ConnectionManager', 'Client state is sent to cluster: ' +
+                        targetClusterId);
+
+                    this.clientState = ClientState.INITIALIZED_ON_CLUSTER;
+                    this.emitLifecycleEvent(LifecycleState.CONNECTED);
+                } else {
+                    this.logger.warn('ConnectionManager', 'Cannot set client state to initialized on ' +
+                        'cluster because current cluster id: ' + this.clusterId + ' is different than expected cluster id: ' +
+                        targetClusterId);
+                }
+            })
+            .catch((error: Error) => {
+                const clusterName = this.client.getConfig().clusterName;
+                this.logger.warn('ConnectionManager', 'Failure during sending state to the cluster: ' +
+                    error.message);
+                if (targetClusterId.equals(this.clusterId)) {
+                    this.logger.warn('ConnectionManager', 'Retrying sending state to the cluster: ' +
+                        targetClusterId + ', name: ' + clusterName);
+                    return this.initializeClientOnCluster(targetClusterId);
+                }
+            });
+    }
+
+    private tryConnectToAllClusterMembers(index: number, members: Member[]): Promise<void> {
+        if (index >= members.length) {
+            return Promise.resolve();
+        }
+
+        const member = members[index];
+        return this.getOrConnect(member.address)
+            .catch(() => {
+                // No-op
+            })
+            .then(() => {
+                return this.tryConnectToAllClusterMembers(index + 1, members);
```

Comment:
It seems that we connect to member sequentially. Wouldn't it be better to use `Promise.all()` here and try to connect to all members simultaneously?

Reply:
You are right. Changed the method as you suggested